source/core/vs_fpcameracontroller.cpp

```diff
@@ -1,218 +1,180 @@
+            cam->setAspectRatio(aspectRatio);
+            inputHandler->frameBufferResizeHandled();
+        }
     }
-    if (zoom >= 45.0F)
+
+    // Calculate mouse coords in world space
     {
-        zoom = 45.0F;
-    }
-    cam->setZoom(zoom);
-}
+        double xPos = inputHandler->getMouseX();
+        double ypos = inputHandler->getMouseY();
+        int width = inputHandler->getDisplayWidth();
+        int height = inputHandler->getDisplayHeight();
+        const auto screenPosCamera = glm::vec3(xPos, double(height) - ypos, 0.F);
+        const auto screenPosFar = glm::vec3(xPos, double(height) - ypos, 1.F);
 
-void VSFPCameraController::processKeyboardInput(GLFWwindow* window, float deltaTime) const
-{
-    float velocity = movementSpeed * deltaTime;
-    glm::vec3 position = cam->getPosition();
+        const auto tmpView = cam->getViewMatrix();
+        const auto tmpProj = cam->getProjectionMatrix();
+        const glm::vec4 viewport = glm::vec4(0.0F, 0.0F, width, height);
 
-    if (glfwGetKey(window, GLFW_KEY_ESCAPE) == GLFW_PRESS)
-    {
-        glfwSetWindowShouldClose(window, GL_TRUE);
-    }
-    if (glfwGetKey(window, GLFW_KEY_W) == GLFW_PRESS)
-    {
-        glm::vec3 front = cam->getFront();
-        position += front * velocity;
-    }
-    if (glfwGetKey(window, GLFW_KEY_S) == GLFW_PRESS)
-    {
-        glm::vec3 front = cam->getFront();
-        position -= front * velocity;
-    }
-    if (glfwGetKey(window, GLFW_KEY_A) == GLFW_PRESS)
-    {
-        glm::vec3 right = cam->getRight();
-        position -= right * velocity;
-    }
-    if (glfwGetKey(window, GLFW_KEY_D) == GLFW_PRESS)
-    {
-        glm::vec3 right = cam->getRight();
-        position += right * velocity;
-    }
-    if (glfwGetKey(window, GLFW_KEY_E) == GLFW_PRESS)
-    {
-        glm::vec3 up = cam->getUp();
-        position += up * velocity;
-    }
-    if (glfwGetKey(window, GLFW_KEY_Q) == GLFW_PRESS)
-    {
-        glm::vec3 up = cam->getUp();
-        position -= up * velocity;
-    }
-    cam->setPosition(position);
-}
+        const auto worldPosNear = glm::unProject(screenPosCamera, tmpView, tmpProj, viewport);
+        const auto worldPosFar = glm::unProject(screenPosFar, tmpView, tmpProj, viewport);
 
-void VSFPCameraController::processFramebufferResize(GLFWwindow* window, int width, int height)
-{
-    (void)window;
+        const auto hitResult = world->getChunkManager()->lineTrace(worldPosNear, worldPosFar);
 
-    float aspectRatio = (float)width / height;
-    cam->setAspectRatio(aspectRatio);
+        if (hitResult.bHasHit)
+        {
+            setMouseInWorldCoords(hitResult.hitLocation);
+            inputHandler->setMouseInWorldPos(hitResult.hitLocation);
+            setMouseNormalInWorldCoords(hitResult.hitNormal);
+
+            world->getDebugDraw()->drawLine(
+                hitResult.hitLocation,
+                hitResult.hitLocation + hitResult.hitNormal * 10.F,
+                {0, 255, 0});
+        }
```

Comment:
While I do agree, that the camera controller is probably not the correct class to calculate and store this information, I do not think the camera is the correct class either. This would mean the camera has to get the cursor position from the input handler, so the overall separation would not increase. Maybe just leave it like this until we find a better solution and once we actually need a trace to somewhere else but the first block that is hit.